### PR TITLE
Add Ruby linter with Steep

### PR DESCRIPTION
**Pull request by lloeki**
Continuation of https://github.com/dense-analysis/ale/pull/4376

Fixes https://github.com/dense-analysis/ale/issues/3254

> Where are the tests? Have you added tests? Have you updated the tests? Read the
comment above and the documentation referenced in it first. Write tests!

> Seriously, read `:help ale-dev` and write tests.

~Working on it (trying to run the test docker thingy).~

Added tests (inspired by Rubocop ones)

### CODE REVIEW

File: ale_linters/ruby/steep.vim [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    \|| a:severity is? 'information'
+    \|| a:severity is? 'hint'
+        return 'W'
+    endif
+
+    return 'E'
+endfunction
+
+" Handle output from steep
+function! ale_linters#ruby#steep#HandleOutput(buffer, lines) abort
```
Comment (w0rp) [inline ale_linters/ruby/steep.vim:61]: This is a complex handler function. Add a test for this too. There's a directory for testing functions for handling results.

**Review by w0rp — CHANGES_REQUESTED** [review]

File: ale_linters/ruby/steep.vim (hunk 1 above)
Comment (lloeki) [reply ale_linters/ruby/steep.vim:61]: Sure, I will!

Comment (lloeki) [reply ale_linters/ruby/steep.vim:61]: I should finally be able to get back to complete this within a few weeks.

Comment (lloeki) [reply ale_linters/ruby/steep.vim:61]: "a few weeks" heh. On it.

Comment (lloeki) [reply ale_linters/ruby/steep.vim:61]: @w0rp I have rebased and added the handler test.

**Comment by lloeki** [comment]
Hmm, test failure on Appveyor because of Windows paths:

```
[00:02:18]   Starting Vader: C:\testplugin\test\linter\test_ruby_steep.vader
[00:02:18]     (1/3) [EXECUTE] Executable should default to steep
[00:02:18]     (1/3) [EXECUTE] (X) ['steep', 'steep check  ''C:\testplugin\test\linter\dummy.rb'''] should be equal to ['steep', 'steep check  ''testplugin/test/linter/dummy.rb''']
[00:02:18]     (2/3) [EXECUTE] Should be able to set a custom executable
[00:02:18]     (2/3) [EXECUTE] (X) ['bin/steep', 'bin/steep check  ''C:\testplugin\test\linter\dummy.rb'''] should be equal to ['bin/steep', 'bin/steep check  ''testplugin/test/linter/dummy.rb''']
[00:02:18]     (3/3) [EXECUTE] Setting bundle appends 'exec steep'
[00:02:18]     (3/3) [EXECUTE] (X) ['path to/bundle', '"path to/bundle" exec steep check  ''C:\testplugin\test\linter\dummy.rb'''] should be equal to ['path to/bundle', '"path to/bundle" exec steep check  ''testplugin/test/linter/dummy.rb''']
[00:02:18]   Success/Total: 0/3
```

**Comment by lloeki** [comment]
Noticed the separator was off for Windows, which would break the replacement and made a fix, but it insists on having `C:` in front still :(

```
[00:02:28]   Starting Vader: C:\testplugin\test\linter\test_ruby_steep.vader
[00:02:28]     (1/3) [EXECUTE] Executable should default to steep
[00:02:28]     (1/3) [EXECUTE] (X) ['steep', 'steep check  ''C:testplugin\test\linter\dummy.rb'''] should be equal to ['steep', 'steep check  ''testplugin/test/linter/dummy.rb''']
[00:02:28]     (2/3) [EXECUTE] Should be able to set a custom executable
[00:02:28]     (2/3) [EXECUTE] (X) ['bin/steep', 'bin/steep check  ''C:testplugin\test\linter\dummy.rb'''] should be equal to ['bin/steep', 'bin/steep check  ''testplugin/test/linter/dummy.rb''']
[00:02:28]     (3/3) [EXECUTE] Setting bundle appends 'exec steep'
[00:02:28]     (3/3) [EXECUTE] (X) ['path to/bundle', '"path to/bundle" exec steep check  ''C:testplugin\test\linter\dummy.rb'''] should be equal to ['path to/bundle', '"path to/bundle" exec steep check  ''testplugin/test/linter/dummy.rb''']
[00:02:28]   Success/Total: 0/3
```

Really though to efficiently debug this as I don't have a Windows setup at hand.

**Comment by lloeki** [comment]
@w0rp there were a few corner-case bugs but it's all fixed and test-covered now. WDYT, good to merge?

**Comment by lloeki** [comment]
Hmm, missed [that one](https://github.com/dense-analysis/ale/actions/runs/9291850874/job/25986245530?pr=4671#step:3:15) locally, will fix:

```
========================================
Running Vint to lint our code
========================================
Vint warnings/errors follow:

ale_linters/ruby/steep.vim:26:21: Use robust operators `==#` or `==?` instead of `==` (see Google VimScript Style Guide (Matching))
ale_linters/ruby/steep.vim:64:19: Use robust operators `==#` or `==?` instead of `==` (see Google VimScript Style Guide (Matching))
```

[Huh what](https://github.com/dense-analysis/ale/actions/runs/9291850874/job/25986245530?pr=4671#step:3:40)?

```
========================================
diff supported-tools.md and doc/ale-supported-languages-and-tools.txt tables
========================================
Differences follow:

--- /tmp/readme.XXMCAcHf
+++ /tmp/doc.XXoaBcKg
@@ -445 +444,0 @@
-Ruby - steep
@@ -446,0 +446 @@
+Ruby - steep
```

**Comment by hsanson** [comment]
Difference here is the sorting order. In supported-tools.md `steep` is listed below `sorbet` while in `ale-supported-languages-and-tools.txt` is listed below standardrb. Make sure is in same exact order on both files.

```
========================================
diff supported-tools.md and doc/ale-supported-languages-and-tools.txt tables
========================================
Differences follow:

--- /tmp/readme.XXMCAcHf
+++ /tmp/doc.XXoaBcKg
@@ -445 +444,0 @@
-Ruby - steep
@@ -446,0 +446 @@
+Ruby - steep
```

**Comment by lloeki** [comment]
Oh my, thanks @hsanson!

Just pushed fixes for the remaining bits. Let's see if CI is green now.

**Review by hsanson — APPROVED** [review]
Thanks for the contribution and patience.
